File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int help(vector<int>& v, int n, int i, vector<int>& dp) {
        if (i >= n) {
            return 0;
        }
        if (dp[i] != -1) {
            return dp[i];
        }
        int take = v[i] + help(v, n, i + 2, dp);
        int not_take = help(v, n, i + 1, dp);
        dp[i] = max(take, not_take);

        return dp[i];
    }
    int rob(vector<int>& v) {
        int n = v.size();
        vector<int> dp(n, -1);
        return help(v, n, 0, dp);
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def helper(self, v: List[int], n: int, dp: List[int]) -> int:
        if n <= 0:
            return 0
        if dp[n] != -1:
            return dp[n]
        take = v[n - 1] + self.helper(v, n - 2, dp)
        skip = self.helper(v, n - 1, dp)
        dp[n] = max(take, skip)
        return dp[n]

    def rob(self, v: List[int]) -> int:
        n = len(v)
        dp = [-1] * (n + 1)
        return self.helper(v, n, dp)
